File: day22/solution.py
#!/usr/bin/env python3
import sys
from functools import cache

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'example.txt'
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg == '-i': FILENAME = 'input.txt'
  elif arg == '-e': FILENAME = 'example.txt'
  elif arg.startswith('-e'): FILENAME = f'example{arg[-1]}.txt'
  elif arg == '-d': DEBUG = 1
  elif arg == '-d2': DEBUG = 2
  elif arg == '-d3': DEBUG = 3
  else: raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  inputs = list(map(int, f.read().splitlines()))


def evolve(value: int, iterations: int) -> int:
  orig_value = value
  for i in range(iterations):
    value ^= value << 6  # xor mult 64
    value &= 16777215  # mod 16777216
    value ^= value >> 5  # xor idiv 32
    value &= 16777215  # mod 16777216
    value ^= value << 11  # xor mult 2048
    value &= 16777215  # mod 16777216
    if DEBUG > 0: print(f'evolve {orig_value} {i+1} {value}')
  return value


def part1():
  # evolve(inputs[0], 10)
  # for value in inputs:
  #   print(f'{value=} {evolve(value,2000)}')
  total=sum( \
    evolve(value,2000) \
    for value in inputs \
  )
  print(f'part 1: {total=}')


part1()

File: day22/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item == '-d' ? 1 : item == '-d2' ? 2 : item == '-d3' ? 3 : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) =>
    item == '-i' ? 'input.txt' : item == '-e' ? 'example.txt' : item.startsWith('-e') ? `example${item.slice(-1)}.txt` : acc,
  'example.txt'
);
console.log({ FILENAME, DEBUG });

const text = await Deno.readTextFile(FILENAME);
if (DEBUG > 1) console.debug({ text });
const codes = text
  .split('\n')
  .filter((line) => line.trim())
  .map((line) => parseInt(line));
if (DEBUG > 1) console.debug({ codes });

const evolve = (value: number) => {
  value ^= value << 6;
  value &= 16777215;
  value ^= value >> 5;
  value &= 16777215;
  value ^= value << 11;
  value &= 16777215;
  return value;
};

const part1 = () => {
  let total = BigInt(0);
  for (let code of codes) {
    for (let i = 0; i < 2000; i++) {
      code = evolve(code);
    }
    total += BigInt(code);
  }
  console.log('part 1:', total);
};

const part2 = () => {
  const totalSequences: Map<string, number> = new Map();
  for (let code of codes) {
    const window = [];
    // per code sequence prices so we can abort if sequence already used
    const codeSequences: Map<string, number> = new Map();
    for (let i = 0; i < 2000; i++) {
      const price = code % 10;
      // no point pruning window on each iteration for only 2000 loops
      window.push(price);
      if (i >= 5) {
        // this is probably the slowest part
        const key = JSON.stringify([
          window[i] - window[i - 1],
          window[i - 1] - window[i - 2],
          window[i - 2] - window[i - 3],
          window[i - 3] - window[i - 4],
        ]);
        // add to codeSequences if not already set
        if (!codeSequences.has(key)) {
          codeSequences.set(key, price);
        }
      }
      code = evolve(code);
    }
    // merge codeSequences into totalSequences
    for (const [key, price] of codeSequences.entries()) {
      totalSequences.set(key, (totalSequences.get(key) ?? 0) + price);
    }
  }
  // select the max value from totalSequences
  const highestSequence = [...totalSequences.entries()].toSorted((a, b) => b[1] - a[1])[0];
  console.log('part 2:', highestSequence);
};

part1();
part2();
